File: Trabajo_9/CPP/Ejercicios/Ejercicio_5.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
using namespace std;

class Producto {
public:
    virtual double calculo() const = 0;
    virtual void aplicarDescuento(double porcentaje) = 0;
    virtual void imprimirDatos() const = 0;
};

class Electronico : public Producto {
private:
    string nombre;
    double precio;
    int cantidad;

public:
    Electronico(string n, double p, int c) {
        this->nombre = n;
        this->precio = p;
        this->cantidad = c;
    }

    double calculo() const override {
        return precio * cantidad;
    }

    void aplicarDescuento(double porcentaje) override {
        precio *= (1 - porcentaje / 100);
    }

    void imprimirDatos() const override {
        cout << "Producto: " << nombre << endl;
        cout << "Categoria: Electronico" << endl;
        cout << "Precio: " << precio << endl;
        cout << "Cantidad: " << cantidad << endl;
    }
};

class Ropa : public Producto {
private:
    string nombre;
    double precio;
    int cantidad;
    string talla;

public:
    Ropa(string n, double p, int c, string t) {
        this->nombre = n;
        this->precio = p;
        this->cantidad = c;
        this->talla = t;
    }

    double calculo() const override {
        return precio * cantidad;
    }

    void aplicarDescuento(double porcentaje) override {
        precio *= (1 - porcentaje / 100);
    }

    void imprimirDatos() const override {
        cout << "Producto: " << nombre << endl;
        cout << "Categoria: Ropa" << endl;
        cout << "Precio: " << precio << endl;
        cout << "Cantidad: " << cantidad << endl;
        cout << "Talla: " << talla << endl;
    }
};

class Alimento : public Producto {
private:
    string nombre;
    double precio;
    int cantidad;
    string vencimiento;

public:
    Alimento(string n, double p, int c, string v) {
        this->nombre = n;
        this->precio = p;
        this->cantidad = c;
        this->vencimiento = v;
    }

    double calculo() const override {
        return precio * cantidad;
    }

    void aplicarDescuento(double porcentaje) override {
        precio *= (1 - porcentaje / 100);
    }

    void imprimirDatos() const override {
        cout << "Producto: " << nombre << endl;
        cout << "Categoria: Alimento" << endl;
        cout << "Precio: " << precio << endl;
        cout << "Cantidad: " << cantidad << endl;
        cout << "Fecha de Expiracion: " << vencimiento << endl;
    }
};

int main() {
    vector<Producto*> inventario;

    inventario.push_back(new Electronico("Laptop", 1500, 10));
    inventario.push_back(new Ropa("Camisa", 30, 50, "S"));
    inventario.push_back(new Alimento("Manzana", 1, 100, "2023-12-31"));

    double total = 0;
    for (Producto* p : inventario) {
        total += p->calculo();
    }

    cout << "Precio fuinal: s/" << total << endl;

    for (Producto* p : inventario) {
        p->aplicarDescuento(10);
    }

    total = 0;
    for (Producto* p : inventario) {
        total += p->calculo();
    }

    cout << "Precio Final: s/" << total << endl;

    cout << "Inventario:" << endl;
    for (Producto* p : inventario) {
        p->imprimirDatos();
        cout << "" << endl;
    }

    for (Producto* p :inventario){
        delete p;
    }

    return 0;
}

File: Trabajo_9/PY/Ejercicios/Ejercicio_5.py
class Producto:
    def __init__(self, nombre, precio, cantidad):
        self.nombre = nombre
        self.precio = precio
        self.cantidad = cantidad

    def valor_inventario(self):
        return self.precio * self.cantidad

    def aplicar_descuento(self, porcentaje):
        self.precio *= (1 - porcentaje)

    def __str__(self):
        return f"{self.nombre}: {self.cantidad} unidades a {self.precio} cada una"
class Electronico(Producto):
    def aplicar_descuento(self, porcentaje):
        porcentaje_extra = 0.05  # 5% de descuento adicional
        super().aplicar_descuento(porcentaje + porcentaje_extra)

class Ropa(Producto):
    def aplicar_descuento(self, porcentaje):
        if porcentaje > 0.2:  # Máximo descuento del 20%
            porcentaje = 0.2
        super().aplicar_descuento(porcentaje)

class Alimento(Producto):
    def valor_inventario(self):
        valor_base = super().valor_inventario()
        return valor_base * 0.9  # Suponiendo una reducción del 10%

    def aplicar_descuento(self, porcentaje):
        porcentaje_extra = 0.1  # 10% de descuento adicional
        super().aplicar_descuento(porcentaje + porcentaje_extra)
# Creación de productos
tv = Electronico("Televisor", 1200, 30)
camisa = Ropa("Camisa", 40, 50)
manzana = Alimento("Manzana", 0.5, 500)

tv.aplicar_descuento(0.1)
camisa.aplicar_descuento(0.25)
manzana.aplicar_descuento(0.05)

print(tv)
print(camisa)
print(manzana)

